compiler/rustc_mir_transform/src/check_finalizers.rs

```diff
@@ -13,6 +13,82 @@ use rustc_trait_selection::infer::TyCtxtInferExt;
 #[derive(PartialEq)]
 pub struct CheckFinalizers;
 
+#[derive(Debug)]
+enum FinalizerErrorKind<'tcx> {
+    /// Does not implement `Send` + `Sync`
+    ThreadUnsafe(Span),
+    /// Does not implement `FinalizerSafe`
+    FinalizerUnsafe(Ty<'tcx>, Span),
+    /// Does not implement `ReferenceFree`
+    ContainsReference,
+    /// Uses a trait object whose concrete type is unknown
+    UnknownTraitObject,
+    /// Calls a function whose definition is unavailable, so we can't be certain it's safe.
+    MissingFnDef,
+}
+
+impl<'tcx> FinalizerErrorKind<'tcx> {
+    fn emit(&self, cx: &FinalizationCtxt<'tcx>) {
+        let arg_sp = cx.tcx.sess.source_map().span_to_snippet(cx.arg).unwrap();
+        let mut err = cx.tcx.sess.psess.dcx.struct_span_err(
+            cx.arg,
+            format!("`{arg_sp}` has a drop method which cannot be safely finalized.",),
```

Comment:
Is that comma at the end valid?

Reply:
The one inside the `format!(..)`? I've just tested without it and it seems it is not necessary (but both compile, so if you want me to push an update I will)